common/io.cpp

```diff
@@ -64,14 +71,89 @@ ExpectedSize ByteWriter::Write(
 	return bytes_to_write;
 }
 
-ExpectedSize StreamWriter::Write(
-	vector<uint8_t>::const_iterator start, vector<uint8_t>::const_iterator end) {
-	os_.write(reinterpret_cast<const char *>(&*start), end - start);
-	if (!os_) {
-		return expected::unexpected(Error(make_error_condition(errc::io_error), ""));
+
+class ReaderStreamBuffer : public streambuf {
+public:
+	ReaderStreamBuffer(Reader &reader) :
+		reader_ {reader},
+		buf_(buf_size_) {};
+	streambuf::int_type underflow() override;
+
+private:
+	static const Vsize buf_size_ = 1024;
+	Reader &reader_;
+	vector<uint8_t> buf_;
+};
+
+streambuf::int_type ReaderStreamBuffer::underflow() {
+	// eback -- pointer to the first char (byte)
+	// gptr  -- pointer to the current char (byte)
+	// egptr -- pointer past the last char (byte)
+
+	// This function is only called if gptr() == nullptr or gptr() >= egptr(),
+	// i.e. if there's nothing more to read.
+	if (this->gptr() >= this->egptr()) {
+		auto ex_n_read = reader_.Read(buf_.begin(), buf_.end());
+		streamsize n_read;
+		if (ex_n_read) {
+			n_read = ex_n_read.value();
+		} else {
+			// There is no way to return an error from underflow(), generally
+			// the streams only care about how much data was read. No data or
+			// less data then requested by the caller of istream.read() means
+			// eofbit and failbit are set. If the user code wants to get the
+			// error or check if there was an error, it needs to check errno.
+			//
+			// So as long as we don't clear errno after a failure in the
+			// reader_.Read() above, error handling works as usual and returning
+			// eof below is all that needs to happen here.
+			//
+			// In case errno is not set for some reason, let's try to get it
+			// from the error with a fallback to a generic I/O error.
+			if (errno == 0) {
```

Comment:
If you test for zero, you should perhaps set it to zero as well, before the call to `Read`. Otherwise it may be that it was non-zero already, but `Read` doesn't use it, and then you end up returning an `errno` which is for an unrelated error. I know that some libc calls explicitly encourage you to set it to zero before calling them.

Reply:
Good point. Although I believe we should (also) set it to zero in our code where we use the syscalls that can set it.

Reply:
OTOH, one would think things like `istream.read()` do `errno = 0` before the call the `read()` syscall.